components/secret-mgt/org.wso2.carbon.identity.secret.mgt.core/src/main/java/org/wso2/carbon/identity/secret/mgt/core/dao/impl/CachedBackedSecretDAO.java

```diff
@@ -0,0 +1,260 @@
+
+        return 2;
+    }
+
+    @Override
+    public Secret getSecretByName(int tenantId, String name) throws SecretManagementException {
+
+        Secret secret = getSecretFromCacheByName(name, tenantId);
+        if (secret != null) {
+            if (log.isDebugEnabled()) {
+                String message = String.format("Cache hit for secret by it's name. Secret name: %s, Tenant id: "
+                        + "%d", name, tenantId);
+                log.debug(message);
+            }
+            return secret;
+        } else {
+            if (log.isDebugEnabled()) {
+                String message = String.format("Cache miss for secret by it's name. Secret name: %s, Tenant id: " +
+                        "%d", name, tenantId);
+                log.debug(message);
+            }
+            secret = secretDAO.getSecretByName(tenantId, name);
+            addSecretToCache(secret);
+        }
+        return secret;
+    }
+
+    @Override
+    public Secret getSecretById(int tenantId, String secretId) throws SecretManagementException {
+
+        Secret secret = getSecretFromCacheById(secretId, MultitenantConstants.SUPER_TENANT_DOMAIN_NAME);
+        if (secret != null) {
+            if (log.isDebugEnabled()) {
+                String message = String.format("Cache hit for secret by it's id. Secret id: %s", secretId);
+                log.debug(message);
+            }
+            return secret;
+        } else {
+            if (log.isDebugEnabled()) {
+                String message = String.format("Cache miss for secret by it's id. Secret id: %s", secretId);
+                log.debug(message);
+            }
+            secret = secretDAO.getSecretById(tenantId, secretId);
+            addSecretToCache(secret);
+        }
+        return secret;
+    }
+
+    @Override
+    public List getSecrets(int tenantId) throws SecretManagementException {
+
+        return secretDAO.getSecrets(tenantId);
+    }
+
+    @Override
+    public void deleteSecretById(int tenantId, String secretId) throws SecretManagementException {
+
+        secretDAO.deleteSecretById(tenantId, secretId);
+        deleteCacheBySecretId(secretId, tenantId);
+    }
+
+    @Override
+    public void deleteSecretByName(int tenantId, String name) throws SecretManagementException {
+
+        secretDAO.deleteSecretByName(tenantId, name);
+        deleteCacheBySecretByName(name, tenantId);
+    }
+
+    @Override
+    public void addSecret(Secret secret) throws SecretManagementException {
+
+        secretDAO.addSecret(secret);
+        addSecretToCache(secret);
+    }
+
+    @Override
+    public void replaceSecret(Secret secret) throws SecretManagementException {
+
+        secretDAO.replaceSecret(secret);
+        deleteSecretFromCache(secret);
+    }
+
+    @Override
+    public boolean isExistingSecret(int tenantId, String secretId) throws SecretManagementException {
+
+        return secretDAO.isExistingSecret(tenantId, secretId);
```

Comment:
We can check the cache, during the isExisting check. You can try to get the entry from cache. if yes, you can return true. Else invoke `secretDAO.isExistingSecret` method. No need to cache the `secretDAO.isExistingSecret` result.

Reply:
resolved